src/platform/datapath_winrdma.c

```diff
@@ -0,0 +1,5151 @@
+        goto ErrorExit;
+    }
+
+    Status = NdspiCreateConnector(
+        NdAdapter,
+        RdmaConnection->OverlappedConnFile,
+        &RdmaConnection->Connector);
+    if (QUIC_FAILED(Status))
+    {
+        QuicTraceLogError(
+            CreateConnectorFailed,
+            "Create Connector failed, status:%d", Status);
+        goto ErrorExit;
+    }
+
+    //
+    // Create Memory Window for the connection
+    //
+    if (RdmaConnection->Flags & RDMA_CONNECTION_FLAG_MEMORY_WINDOW_USED)
+    {
+        Status = NdspiCreateMemoryWindow(NdAdapter, &RdmaConnection->RecvMemoryWindow);
+        if (QUIC_FAILED(Status))
+        {
+            QuicTraceLogError(
+                CreateRecvMemoryWindowFailed,
+                "Create RecvMemoryWindow failed, status:%d", Status);
+            goto ErrorExit;
+        }
+        
+        if (RdmaConnection->Flags & RDMA_CONNECTION_FLAG_OFFSET_BUFFER_USED)
+        {
+            Status = NdspiCreateMemoryWindow(NdAdapter, &RdmaConnection->OffsetMemoryWindow);
+            if (QUIC_FAILED(Status))
+            {
+                QuicTraceLogError(
+                    CreateSendMemoryWindowFailed,
+                    "Create SendMemoryWindow failed, status:%d", Status);
+                goto ErrorExit;
+            }
+        }
+    }
+    else
+    {
+        RdmaConnection->RecvMemoryWindow = NULL;
+        RdmaConnection->OffsetMemoryWindow = NULL;
+    }
+
+    SocketProc->RdmaHandle = RdmaConnection->OverlappedConnFile;
+
+    //
+    // Disable automatic IO completions being queued if the call completes
+    // synchronously. This is because we want to be able to complete sends
+    // inline, if possible.
+    //
+    if (!SetFileCompletionNotificationModes(
+        SocketProc->RdmaHandle,
+        FILE_SKIP_COMPLETION_PORT_ON_SUCCESS | FILE_SKIP_SET_EVENT_ON_HANDLE))
+    {
+        DWORD LastError = GetLastError();
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Socket,
+            LastError,
+            "SetFileCompletionNotificationModes");
+        Status = HRESULT_FROM_WIN32(LastError);
+        goto ErrorExit;
+    }
+
+    //
+    // Finally Create Socket object
+    //
+    if (Type != CXPLAT_SOCKET_RDMA_SERVER)
+    {
+        SocketProc->DatapathProc = &Datapath->Partitions[PartitionIndex];
+        CxPlatRefIncrement(&SocketProc->DatapathProc->RefCount);
+
+        if (!CxPlatEventQAssociateHandle(
+            SocketProc->DatapathProc->EventQ,
+            SocketProc->RdmaHandle)) 
+        {
+            DWORD LastError = GetLastError();
+            QuicTraceEvent(
+                DatapathErrorStatus,
+                "[data][%p] ERROR, %u, %s.",
+                Socket,
+                LastError,
+                "CreateIoCompletionPort");
+            Status = HRESULT_FROM_WIN32(LastError);
+            goto ErrorExit;
+        }
+
+        if (RemoteAddress != NULL)
```

Comment:
What does it mean for a client socket to not have a remote address? Is it possible / valid?